### PR TITLE
Implemented new modifiers for mvaTTH Run 3

### PR DESCRIPTION
**Pull request by Cvico**
## Brief description
Dear reviewers,

This PR is to introduce the newest training of the Muon mvaTTH in CMSSW. The training has been performed following what was done already done for Run 2, but using Run 3 MC samples for training and validation.

## Context of the PR
The goal for this new training is to be used as a substitute of the current mvaTTH **only in Run 3 MC production**. That is: for Run 2 we still need to keep the current mvaTTH implementation (can be found in these lines), and for Run 3 we want to make use of an era modifier to fetch the proper --new -- weights that have been computed.

## What this PR includes
**This change should only affect Run 3 production**. For that, this PR includes:
 1. The required era modifiers in `Configuration/Eras/`.
     * I've added two new era modifiers `run3_muon_2022` and `run3_muon_2023`, which are called from the general `Run3_Eras_cff.py`. 
     * I hope this is the optimal way to handle our use case.
 3. The call to the era modifier and the implementation of the new mvaTTH evaluation.
     * How the mva is evaluated changes a little bit with respect to the previous version: mainly the order in which the variables are parsed to TMVA and also we have added a new additional variable which is `LepGood_pfRelIso03_all`.
     * Which weights should be read in order to properly compute the mva score.

## Validation
I've performed a local test running a simple `cmsRun` script to see that the mva is loaded and nanoAOD is in fact produced. 

## Other comments
Please let me know if the proposal is reasonable, and if there's anything else I can do for testing. I'm adding the L2 muon convener @JanFSchulte so he can follow up the PR as well. Feel free to add any other muon convener (I don't have their github user names unfortunately).

## Linked PRs
This is the link for the PR that includes the weights in CMS-data: [pull16](https://github.com/cms-data/PhysicsTools-NanoAOD/pull/16).

### CODE REVIEW

**Comment by vlimant** [comment]
I'd rather have the default be run3, and you revert to the old using `~run3`

**Comment by Cvico** [comment]
> I'd rather have the default be run3, and you revert to the old using `~run3`

Thanks @vlimant for the suggestion. I don't have a feeling on what's the best option here, but could you point me on how to properly do this? 

Do I have to change [these lines](https://github.com/cms-sw/cmssw/blob/ba550c090f5a34bf49ecfda0430eb485aaa8d885/PhysicsTools/NanoAOD/python/muons_cff.py#L89-L110), and call the run2 modifier instead of the run3 modifier?

File: PhysicsTools/NanoAOD/python/muons_cff.py [hunk 1]
```diff
@@ -119,6 +119,30 @@
     weightFile = "PhysicsTools/NanoAOD/data/mu_BDTG_2016.weights.xml",
 )
 
+run3_muon.toModify(
```
Comment (vlimant) [inline PhysicsTools/NanoAOD/python/muons_cff.py:122]: in line 89, set `muonMVATTH` with the run3 parameters, and there do `run3_common).toModify(` with old paremters. which is just changing the fileinpath, the variable order and remove variables

Comment (Cvico) [reply PhysicsTools/NanoAOD/python/muons_cff.py]: I see, I've updated the config to do this. I was not aware this `~run3` could be done. Let me know if it is what you expected.

**Comment by JanFSchulte** [comment]
Hi @Cvico From the POG side we would like to take the opportunity to rename the variable from mvaTTH to mvaPrompt to unify the naming with what is used in the MUO-22-001 paper and to not have it so closely associated with just one analysis since we want this to be an official POG product going forward. Could  you take care of that? It should probably be mentioned in the doc string what the previous name was.

**Comment by Cvico** [comment]
> Hi @Cvico From the POG side we would like to take the opportunity to rename the variable from mvaTTH to mvaPrompt to unify the naming with what is used in the MUO-22-001 paper and to not have it so closely associated with just one analysis since we want this to be an official POG product going forward. Could you take care of that? It should probably be mentioned in the doc string what the previous name was.

Hi Jan. From my side I personally like the idea of changing the name to mvaPrompt (more precisely, I don't have strong opinion against it). But since there's a mvaTTH for electrons, I think EGM conveners should be aware of the change as well. I can quickly contact them with the proposal.

**Comment by Cvico** [comment]
Hi, I've updated the name from  `mvaTTH`  to `promptMVA` as @JanFSchulte suggested, also in the `electrons_cff.py` config. I've checked it produces an output with two branches: `Electron_promptMVA` and `Muon_promptMVA`. 

I've also included two changes: one in `nano_eras_cff.py ` to include the `run3_muon` modifiers, and also I had to modify the muonMVALowPt to fetch the variables for the Run 2 training explicitly. The reason for this latter change is because it was previously working on TOP of the Run 2 mvaTTH configuration, but since we moved this to Run 3 by default, the mvaLowPT had to be explicitly modified with the Run 2 variables.

**Comment by RSalvatico** [comment]
type egamma

File: PhysicsTools/NanoAOD/python/electrons_cff.py [hunk 2]
```diff
@@ -359,7 +359,7 @@ def _get_bitmapVIDForEle_docstring(modules,WorkingPoints):
         jetNDauCharged = Var("?userCand('jetForLepJetVar').isNonnull()?userFloat('jetNDauChargedMVASel'):0", "uint8", doc="number of charged daughters of the closest jet"),
     ),
     externalVariables = cms.PSet(
-        mvaTTH = ExtVar(cms.InputTag("electronMVATTH"),float, doc="TTH MVA lepton ID score",precision=14),
+        promptMVA = ExtVar(cms.InputTag("muonPROMPTMVA"),float, doc="Prompt MVA lepton ID score. Corresponds to the previous mvaTTH",precision=14),
```
Comment (AlbertoBelvedere) [inline PhysicsTools/NanoAOD/python/electrons_cff.py:362]: Hello! I think here you forgot to change the name from muonPROMPTMVA to electronPROMPTMVA

File: PhysicsTools/NanoAOD/python/muons_cff.py [hunk 3]
```diff
@@ -86,39 +86,76 @@
...
+        LepGood_dz = cms.string("log(abs(dB('PVDZ')))"),
+        LepGood_segmentComp = cms.string("segmentCompatibility"),
+    )
 )
 
 run2_muon_2016.toModify(
-    muonMVATTH,
+    muonPROMPTMVA,
     weightFile = "PhysicsTools/NanoAOD/data/mu_BDTG_2016.weights.xml",
 )
```
Comment (makortel) [inline PhysicsTools/NanoAOD/python/muons_cff.py:136]: Given the `~run3_muon` customization below, this `run2_muon_2016` customization is now ineffective in practice.

**Comment by Cvico** [comment]
Hi, comments addressed. I also spotted an issue in the Era Run3 modifier: since run2_muon_2018 was not being explicitly removed there, it was loading the Run 2 weights anyway... Should I also put run2_muon_2017 in that exclusion file? I think now it should be only reading 2022 when the Era config is not run2_muon_2016, 2017 or 2018... But following all this modifiers is a little bit complicated.

**Comment by makortel** [comment]
> I also spotted an issue in the Era Run3 modifier: since run2_muon_2018 was not being explicitly removed there, it was loading the Run 2 weights anyway...

By quick look excluding `run2_muon_2018` modifier from `Run3` era would be consistent with how `run2_muon201[67]` modifiers are handled in `Run2_201[78]` eras. That change would seem to impact only two places.


----

First case is
https://github.com/cms-sw/cmssw/blob/2e5dc6dfc7a5b5c7fc69b46f79859ee02b97e3b3/PhysicsTools/NanoAOD/python/triggerObjects_cff.py#L390-L395


AFAICT the `prefiringweight` module is enabled only with
https://github.com/cms-sw/cmssw/blob/2e5dc6dfc7a5b5c7fc69b46f79859ee02b97e3b3/PhysicsTools/NanoAOD/python/triggerObjects_cff.py#L422-L424
and `_run2_HLTconditions` is specific to the eras `Run2_201[678]`, so the change would be safe.

----

In the second case
https://github.com/cms-sw/cmssw/blob/2e5dc6dfc7a5b5c7fc69b46f79859ee02b97e3b3/PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py#L42
the question is whether these values should apply to `Run3` as well.

----

(on a side note, completely irrelevant to this PR, this customization
https://github.com/cms-sw/cmssw/blob/2e5dc6dfc7a5b5c7fc69b46f79859ee02b97e3b3/PhysicsTools/NanoAOD/python/triggerObjects_cff.py#L372-L375
violates 6.16 in https://cms-sw.github.io/cms_coding_rules.html#6--packaging-rules-1)

**Comment by makortel** [comment]
> Should I also put run2_muon_2017 in that exclusion file?

In what exclusion file?

File: PhysicsTools/NanoAOD/python/muons_cff.py (hunk 3 above)
Comment (makortel) [reply PhysicsTools/NanoAOD/python/muons_cff.py:136]: In the present setup this `run2_muon_2016` customization is applied to the `muonPROMPTMVA` that IIUC corresponds the Run 3 setup (that has e.g. `LepGood_pfRelIso03_all` variable that was not present in the earlier Run2-based base). Is that the intention?

**Comment by Cvico** [comment]
> > Should I also put run2_muon_2017 in that exclusion file?
> 
> In what exclusion file?

I was refering to the `Eras_Run3_cff.py` file where I explicitly excluded the 2018_muon config. But I think inside the 2018_muon config one removes the 2017_muon config, so I don't see need for removing it again

**Comment by Cvico** [comment]
> > I also spotted an issue in the Era Run3 modifier: since run2_muon_2018 was not being explicitly removed there, it was loading the Run 2 weights anyway...
> 
> By quick look excluding `run2_muon_2018` modifier from `Run3` era would be consistent with how `run2_muon201[67]` modifiers are handled in `Run2_201[78]` eras. That change would seem to impact only two places.
> 
> First case is
> 
> https://github.com/cms-sw/cmssw/blob/2e5dc6dfc7a5b5c7fc69b46f79859ee02b97e3b3/PhysicsTools/NanoAOD/python/triggerObjects_cff.py#L390-L395
> 
> AFAICT the `prefiringweight` module is enabled only with
> 
> https://github.com/cms-sw/cmssw/blob/2e5dc6dfc7a5b5c7fc69b46f79859ee02b97e3b3/PhysicsTools/NanoAOD/python/triggerObjects_cff.py#L422-L424
> 
> and `_run2_HLTconditions` is specific to the eras `Run2_201[678]`, so the change would be safe.
> 
> In the second case
> 
> https://github.com/cms-sw/cmssw/blob/2e5dc6dfc7a5b5c7fc69b46f79859ee02b97e3b3/PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py#L42
> 
> the question is whether these values should apply to `Run3` as well.
> 
> (on a side note, completely irrelevant to this PR, this customization
> 
> https://github.com/cms-sw/cmssw/blob/2e5dc6dfc7a5b5c7fc69b46f79859ee02b97e3b3/PhysicsTools/NanoAOD/python/triggerObjects_cff.py#L372-L375
> 
> violates 6.16 in https://cms-sw.github.io/cms_coding_rules.html#6--packaging-rules-1)

I think @JanFSchulte @Fedespring might comment better on whether these things are needed for Run 3

**Comment by makortel** [comment]
> > > Should I also put run2_muon_2017 in that exclusion file?
> > 
> > 
> > In what exclusion file?
> 
> I was refering to the `Eras_Run3_cff.py` file where I explicitly excluded the 2018_muon config. But I think inside the 2018_muon config one removes the 2017_muon config, so I don't see need for removing it again

Right. The `run2_muon_2017` is already excluded in
https://github.com/cms-sw/cmssw/blob/2e5dc6dfc7a5b5c7fc69b46f79859ee02b97e3b3/Configuration/Eras/python/Era_Run2_2018_cff.py#L24
so excluding it again in something deriving from `Run2_2018` would be confusing.

File: Configuration/Eras/python/Modifier_run3_muon_cff.py [hunk 4]
```diff
@@ -0,0 +1,3 @@
+import FWCore.ParameterSet.Config as cms
+
+run3_muon=cms.Modifier()
```
Comment (vlimant) [inline Configuration/Eras/python/Modifier_run3_muon_cff.py:3]: is it still needed in the current implementation ? I don't think so.

File: PhysicsTools/NanoAOD/python/muons_cff.py [hunk 5]
```diff
@@ -86,39 +86,76 @@
...
         LepGood_segmentComp = cms.string("segmentCompatibility"),
 
     )
 )
 
-muonMVALowPt = muonMVATTH.clone(
+muonMVALowPt = muonPROMPTMVA.clone(
     weightFile =  cms.FileInPath("PhysicsTools/NanoAOD/data/mu_BDTG_lowpt.weights.xml"),
     name = cms.string("muonMVALowPt"),
+    variablesOrder = cms.vstring(["LepGood_pt","LepGood_eta","LepGood_jetNDauChargedMVASel","LepGood_miniRelIsoCharged","LepGood_miniRelIsoNeutral","LepGood_jetPtRelv2","LepGood_jetDF","LepGood_jetPtRatio","LepGood_dxy","LepGood_sip3d","LepGood_dz","LepGood_segmentComp"]),
```
Comment (vlimant) [inline PhysicsTools/NanoAOD/python/muons_cff.py:116]: how about writing the `variables` and the `variableOrder` only once.
`_muonVariablesOrderLegacy` to be used for `muonMVALowPt` and in the `run2_muon_*` modifiers.
`_muonVariablesLegacy` (with the common variables) `_muonVariablesExt` (with the extra ones ; use `_muonVariablesLegacy+ _muonVariablesExt` in run3 (default config) and the `_muonVariablesLegacy` in the modifiers

**Comment by JanFSchulte** [comment]
> > > I also spotted an issue in the Era Run3 modifier: since run2_muon_2018 was not being explicitly removed there, it was loading the Run 2 weights anyway...
> > 
> > 
> > By quick look excluding `run2_muon_2018` modifier from `Run3` era would be consistent with how `run2_muon201[67]` modifiers are handled in `Run2_201[78]` eras. That change would seem to impact only two places.
> > First case is
> > https://github.com/cms-sw/cmssw/blob/2e5dc6dfc7a5b5c7fc69b46f79859ee02b97e3b3/PhysicsTools/NanoAOD/python/triggerObjects_cff.py#L390-L395
> > 
> > AFAICT the `prefiringweight` module is enabled only with
> > https://github.com/cms-sw/cmssw/blob/2e5dc6dfc7a5b5c7fc69b46f79859ee02b97e3b3/PhysicsTools/NanoAOD/python/triggerObjects_cff.py#L422-L424
> > 
> > and `_run2_HLTconditions` is specific to the eras `Run2_201[678]`, so the change would be safe.
> > In the second case
> > https://github.com/cms-sw/cmssw/blob/2e5dc6dfc7a5b5c7fc69b46f79859ee02b97e3b3/PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py#L42
> > 
> > the question is whether these values should apply to `Run3` as well.
> > (on a side note, completely irrelevant to this PR, this customization
> > https://github.com/cms-sw/cmssw/blob/2e5dc6dfc7a5b5c7fc69b46f79859ee02b97e3b3/PhysicsTools/NanoAOD/python/triggerObjects_cff.py#L372-L375
> > 
> > violates 6.16 in https://cms-sw.github.io/cms_coding_rules.html#6--packaging-rules-1)
> 
> I think @JanFSchulte @Fedespring might comment better on whether these things are needed for Run 3

For the L1 prefiring corrections, we do not have any new weights for Run 3 (yet), those would be in the work in the L1 group. So excluding the relevant modifications from Run 3 would make sense, since the values are meaningless for Run 3 anyway. I'm not sure what the dummy values filled in that case are, though. 

As for the effective areas for MiniISO, there is no retuning for Run 3, so we would need them to stay as they are for 2018 in all Run 3 Mini/Nano versions.

File: PhysicsTools/NanoAOD/python/muons_cff.py (hunk 5 above)
Comment (Cvico) [reply PhysicsTools/NanoAOD/python/muons_cff.py:116]: You mean to add this inside the `muons_cff.py` script directly? Or add this in the run2_muon modifiers?

**Comment by vlimant** [comment]
NB #44392

File: PhysicsTools/NanoAOD/python/muons_cff.py (hunk 5 above)
Comment (vlimant) [reply PhysicsTools/NanoAOD/python/muons_cff.py:116]: hard to write down what I meant ; in essence with the least re-writing of the same string is preferred. please note #44392 that I created in a view to streamline that module configuration

**Comment by makortel** [comment]
> As for the effective areas for MiniISO, there is no retuning for Run 3, so we would need them to stay as they are for 2018 in all Run 3 Mini/Nano versions.

Then, given that the present state of this PR excludes the `run2_muon_2018` modifier from the `Run3` era, the
https://github.com/cms-sw/cmssw/blob/2e5dc6dfc7a5b5c7fc69b46f79859ee02b97e3b3/PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py#L40-L42
needs a new line
```py
run3_muon.toModify( process.patMuons, effectiveAreaVec = [0.0566, 0.0562, 0.0363, 0.0119, 0.0064])
```
or you could consolidate
```py
(run2_muon_2017 | run2_muon_2018 | run3_muon).toModify( process.patMuons, effectiveAreaVec = [0.0566, 0.0562, 0.0363, 0.0119, 0.0064])
```

File: Configuration/Eras/python/Modifier_run3_muon_cff.py (hunk 4 above)
Comment (Cvico) [reply Configuration/Eras/python/Modifier_run3_muon_cff.py:3]: I think for [this comment](https://github.com/cms-sw/cmssw/pull/44322#issuecomment-1994508415), run3_muon is actually useful. So I'd be inclined to keep it.

**Comment by Cvico** [comment]
> > As for the effective areas for MiniISO, there is no retuning for Run 3, so we would need them to stay as they are for 2018 in all Run 3 Mini/Nano versions.
> 
> Then, given that the present state of this PR excludes the `run2_muon_2018` modifier from the `Run3` era, the
> 
> https://github.com/cms-sw/cmssw/blob/2e5dc6dfc7a5b5c7fc69b46f79859ee02b97e3b3/PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py#L40-L42
> 
> 
> needs a new line
> ```python
> run3_muon.toModify( process.patMuons, effectiveAreaVec = [0.0566, 0.0562, 0.0363, 0.0119, 0.0064])
> ```
> 
> or you could consolidate
> 
> ```python
> (run2_muon_2017 | run2_muon_2018 | run3_muon).toModify( process.patMuons, effectiveAreaVec = [0.0566, 0.0562, 0.0363, 0.0119, 0.0064])
> ```

Fixed. I opted for the first option.

**Comment by jfernan2** [comment]
please test

**Comment by Cvico** [comment]
Hi, I fixed the import that failed. Apologies for that rookie mistake.

**Comment by Cvico** [comment]
Hi, I see #44392 has been merged. How do I properly resolve the conflicts? Do I pull the commit and then redo my changes from there?

**Comment by vlimant** [comment]
hello, the only conflict is in muon_cff, and the change to BaseMVAValueMapProducer might actually help with this PR. 
the "fun" will begin for the backport, if backport at all, since it is going to be different. I don't think I want just yet to backport #44392

**Comment by JanFSchulte** [comment]
If at all possible, we would really like to have this back-ported to 14_0_X so we can have the new training in the Nano for 2024 data and the 2022/2023 ReReco.

**Comment by JanFSchulte** [comment]
Ah, and @Cvico, the cleanest solution would be to rebase your branch to the latest IB, following https://cms-sw.github.io/tutorial-resolve-conflicts.html

**Comment by Cvico** [comment]
Apologies for the low experience on rebasing:
What should be the new base? I'm doing `git cms-rebase-topic cvico:mvatth_run3_14_1_0_pre0` from `14_1_0_pre0` but it does not find any conflict...

**Comment by JanFSchulte** [comment]
You need to rebase with a CMSSW build that has the PR that created the conflict already compiled in. There are twice-daily integration builds that reflect the current state of the repository. In this case, you need to start from a fresh checkout of CMSSW_14_1_X_2024-03-19-1100 and then rebase your branch on that.

**Comment by vlimant** [comment]
Hi @Cvico , if you have negative experience with rebasing, as this can get complicated (but a good experience for sft dev), I can take care of cherrypicking your changes and do the necessary in muons_cff in a new PR ; let me know how it goes on your end.
In the meatime, I'll backport #44392 too, to make it easier to get your feature in 14.0.

**Comment by vlimant** [comment]
I made #44503 to simplify this.

**Comment by vlimant** [comment]
pleas close

**Comment by Cvico** [comment]
> Hi @Cvico , if you have negative experience with rebasing, as this can get complicated (but a good experience for sft dev), I can take care of cherrypicking your changes and do the necessary in muons_cff in a new PR ; let me know how it goes on your end. In the meatime, I'll backport #44392 too, to make it easier to get your feature in 14.0.

Hi @vlimant, sorry for the late reply, yesterday I was busy with teaching. I'm not sure what's the best option here, I have no experience with it, I'm not sure if (for time sake) it would be better to fix in another PR.

**Comment by vlimant** [comment]
please close

**Pull request closed without merging**